ovos_bus_client/users.py

```diff
@@ -0,0 +1,223 @@
+            if self._state:
+                self._address += ", " + self._state
+            if self._country:
+                self._address += ", " + self._country
+        return self._address
+
+    @property
+    def city(self):
+        if not self._city:
+            self._reverse_geolocate()
+        return self._city
+
+    @property
+    def state(self):
+        if not self._state:
+            self._reverse_geolocate()
+        return self._state
+
+    @property
+    def country(self):
+        if not self._country:
+            self._reverse_geolocate()
+        return self._country
+
+    @property
+    def country_code(self):
+        if not self._country_code:
+            self._reverse_geolocate()
+        return self._country_code
+
+    @classmethod
+    def _reverse_geolocate(cls):
+        data = cls.geolocation.get_reverse_geolocation(lat, lon)
+        loc = cls.from_dict(data)
+        return loc
+
+    @classmethod
+    def from_address(cls, address):
+        data = cls.geolocation.get_geolocation(address) or {}
+        data["address"] = address
+        location = cls.from_dict(data)
+        return location
+
+    @classmethod
+    def from_dict(cls, data):
+        if isinstance(data, str):
+            data = json.loads(data)
+
+        latitude = data["coordinate"]["latitude"]
+        longitude = data["coordinate"]["longitude"]
+        loc = Location(latitude, longitude)
+        # avoid geolocation api calls if possible
+        loc._city = data["city"]["name"]
+        loc._state = data["city"]["state"].get("name")
+        loc._country = data["city"]["state"]["country"]["name"]
+        loc._timezone_code = data.get("timezone", {}).get("name")
+        loc._address = data.get("address")
+        return loc
+
+    def as_dict(self):
+        return {
+            "city": {
+                "name": self.city,
+                "state": {
+                    "name": self.state,
+                    "country": {
+                        "code": self.country_code,
+                        "name": self.country
+                    }
+                }
+            },
+            "coordinate": {
+                "latitude": self.latitude,
+                "longitude": self.longitude
+            },
+            "timezone": {
+                "name": self.timezone_code
+            },
+            "address": self.address
+        }
+
+
+class UserData:
+    """
+    An class representing a user, this session data comes in message.context
+    It may be provided by a OVOS plugin along the stack
+    (eg, speaker recognition) or from clients (eg, a chat application)
+    """
+
+    def __init__(self, user_id=None, session=None,
+                 preferences=None, location=None):
+        self.user_id = user_id or str(uuid4())
+        self.session = session or SessionManager.get()
+        self.preferences = _config2preferences(preferences)  # equivalent to a mycroft.conf per user
```

Comment:
maybe name this `core_config` to match other common references so its obvious what config this is

Reply:
_config2preferences is meant to filter keys. only a subset will be exposed to users